File: iOS-SDK/PhoneAgentDemo/SDK/Agent/AgentResult.swift
//
//  AgentResult.swift
//  PhoneAgentSDK
//
//  Created by AutoGLM on 2025-01-30.
//

import Foundation

/// Agent 执行结果
public struct AgentResult {

    /// 是否成功完成
    public let success: Bool

    /// 最终消息
    public let message: String?

    /// 执行的总步数
    public let totalSteps: Int

    /// 是否因达到最大步数而停止
    public let reachedMaxSteps: Bool

    /// 执行历史（每一步的详情）
    public let history: [StepInfo]

    /// 性能统计
    public let metrics: ExecutionMetrics?

    public struct StepInfo {
        /// 步骤编号
        public let step: Int

        /// 模型的思考过程
        public let thinking: String

        /// 执行的动作
        public let action: String

        /// 动作类型
        public let actionType: ModelResponse.ActionType

        /// 是否成功
        public let success: Bool

        /// 步骤消息
        public let message: String?

        /// 该步的耗时
        public let duration: TimeInterval
    }

    public struct ExecutionMetrics {
        /// 总耗时
        public let totalDuration: TimeInterval

        /// 平均每步耗时
        public let averageStepDuration: TimeInterval

        /// 模型推理总耗时
        public let totalInferenceTime: TimeInterval

        /// 操作执行总耗时
        public let totalExecutionTime: TimeInterval
    }

    public init(
        success: Bool,
        message: String? = nil,
        totalSteps: Int,
        reachedMaxSteps: Bool = false,
        history: [StepInfo] = [],
        metrics: ExecutionMetrics? = nil
    ) {
        self.success = success
        self.message = message
        self.totalSteps = totalSteps
        self.reachedMaxSteps = reachedMaxSteps
        self.history = history
        self.metrics = metrics
    }
}

// MARK: - 便捷初始化
extension AgentResult {

    /// 成功结果
    public static func success(message: String, totalSteps: Int, history: [StepInfo] = []) -> AgentResult {
        return AgentResult(
            success: true,
            message: message,
            totalSteps: totalSteps,
            history: history
        )
    }

    /// 失败结果
    public static func failure(message: String, totalSteps: Int, history: [StepInfo] = []) -> AgentResult {
        return AgentResult(
            success: false,
            message: message,
            totalSteps: totalSteps,
            history: history
        )
    }

    /// 达到最大步数
    public static func maxStepsReached(totalSteps: Int, history: [StepInfo] = []) -> AgentResult {
        return AgentResult(
            success: false,
            message: "达到最大执行步数",
            totalSteps: totalSteps,
            reachedMaxSteps: true,
            history: history
        )
    }
}

File: Android-SDK/phoneagent-sdk/src/main/java/com/autoglm/phoneagent/agent/AgentResult.kt
package com.autoglm.phoneagent.agent

import com.autoglm.phoneagent.action.ActionType

/**
 * Result of an agent execution.
 *
 * @property success Whether the task completed successfully
 * @property message Optional message describing the final result
 * @property totalSteps Total number of steps executed
 * @property reachedMaxSteps Whether execution stopped because max steps was reached
 * @property history List of step information for each executed step
 * @property metrics Optional performance metrics
 */
data class AgentResult(
    val success: Boolean,
    val message: String? = null,
    val totalSteps: Int = 0,
    val reachedMaxSteps: Boolean = false,
    val history: List<StepInfo> = emptyList(),
    val metrics: ExecutionMetrics? = null
) {
    /**
     * Information about a single execution step.
     *
     * @property step Step number (1-indexed)
     * @property thinking The model's thinking process for this step
     * @property action The action string that was executed
     * @property actionType The parsed action type
     * @property success Whether this step executed successfully
     * @property message Optional message from step execution
     * @property duration Time taken for this step in milliseconds
     * @property inferenceTime Time spent on model inference
     * @property executionTime Time spent executing the action
     */
    data class StepInfo(
        val step: Int,
        val thinking: String,
        val action: String,
        val actionType: ActionType,
        val success: Boolean,
        val message: String? = null,
        val duration: Long = 0,
        val inferenceTime: Long = 0,
        val executionTime: Long = 0
    )

    /**
     * Performance metrics for the entire execution.
     *
     * @property totalDuration Total execution time in milliseconds
     * @property averageStepDuration Average time per step in milliseconds
     * @property totalInferenceTime Total time spent on model inference
     * @property totalExecutionTime Total time spent executing actions
     */
    data class ExecutionMetrics(
        val totalDuration: Long,
        val averageStepDuration: Double,
        val totalInferenceTime: Long,
        val totalExecutionTime: Long
    )

    companion object {
        /**
         * Create a successful result.
         */
        fun success(
            message: String? = null,
            totalSteps: Int = 0,
            history: List<StepInfo> = emptyList(),
            metrics: ExecutionMetrics? = null
        ): AgentResult {
            return AgentResult(
                success = true,
                message = message,
                totalSteps = totalSteps,
                reachedMaxSteps = false,
                history = history,
                metrics = metrics
            )
        }

        /**
         * Create a failed result.
         */
        fun failure(
            message: String,
            totalSteps: Int = 0,
            history: List<StepInfo> = emptyList(),
            metrics: ExecutionMetrics? = null
        ): AgentResult {
            return AgentResult(
                success = false,
                message = message,
                totalSteps = totalSteps,
                reachedMaxSteps = false,
                history = history,
                metrics = metrics
            )
        }

        /**
         * Create a result indicating max steps was reached.
         */
        fun maxStepsReached(
            totalSteps: Int,
            history: List<StepInfo> = emptyList(),
            metrics: ExecutionMetrics? = null
        ): AgentResult {
            return AgentResult(
                success = false,
                message = "Reached maximum steps ($totalSteps) without completion",
                totalSteps = totalSteps,
                reachedMaxSteps = true,
                history = history,
                metrics = metrics
            )
        }

        /**
         * Create a result indicating too many unknown actions.
         */
        fun tooManyUnknownActions(
            totalSteps: Int,
            history: List<StepInfo> = emptyList(),
            metrics: ExecutionMetrics? = null
        ): AgentResult {
            return AgentResult(
                success = false,
                message = "Too many unknown actions detected",
                totalSteps = totalSteps,
                reachedMaxSteps = false,
                history = history,
                metrics = metrics
            )
        }

        /**
         * Create a result indicating user takeover.
         */
        fun userTakeover(
            message: String = "User took over the task",
            totalSteps: Int = 0,
            history: List<StepInfo> = emptyList(),
            metrics: ExecutionMetrics? = null
        ): AgentResult {
            return AgentResult(
                success = false,
                message = message,
                totalSteps = totalSteps,
                reachedMaxSteps = false,
                history = history,
                metrics = metrics
            )
        }
    }

    /**
     * Get a human-readable summary of the result.
     */
    fun getSummary(): String {
        return when {
            success -> "✓ 任务完成${if (!message.isNullOrBlank()) ": $message" else ""}"
            reachedMaxSteps -> "✗ 达到最大步数限制 ($totalSteps 步)"
            else -> "✗ 任务失败${if (!message.isNullOrBlank()) ": $message" else ""}"
        }
    }

    /**
     * Get detailed information about each step.
     */
    fun getDetailedHistory(): String {
        if (history.isEmpty()) return "无执行记录"

        return buildString {
            appendLine("=== 执行历史 ===")
            history.forEach { step ->
                appendLine("步骤 ${step.step}:")
                appendLine("  思考: ${step.thinking}")
                appendLine("  动作: ${step.action}")
                appendLine("  类型: ${step.actionType}")
                appendLine("  状态: ${if (step.success) "✓ 成功" else "✗ 失败"}")
                if (step.message != null) {
                    appendLine("  消息: ${step.message}")
                }
                appendLine("  耗时: ${step.duration}ms")
                appendLine()
            }
        }
    }
}
